text/0000-bench-utils.md

```diff
@@ -0,0 +1,208 @@
+trigger in practice. Another application is writing constant time code, where it
+is undesirable for the compiler to optimize certain operations depending on the
+context in which they are executed.
+
+The implementation of this function is backend-specific and currently requires
+inline assembly. No viable alternative is available in stable Rust.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+## `hint::black_box`
+
+The function:
+
+```rust
+/// An _unknown_ function that returns `x`.
+pub fn black_box<T>(x: T) -> T;
+```
+
+is an _unknown_ function, that is, a function that the compiler cannot make any
+assumptions about. It can use `x` in any possible valid way that Rust code is
+allowed to without introducing undefined behavior in the calling code. This
+requires the compiler to be maximally pessimistic in terms of optimizations, but
+the compiler is still allowed to optimize the expression generating `x`. This
+function returns `x` and is a no-op in the virtual machine.
+
+For example ([`rust.godbolt.org`](https://godbolt.org/g/YP2GCJ)):
+
+```rust
+fn foo(x: i32) -> i32{ 
+  hint::black_box(2 + x);
+  3
+}
+let a = foo(2);
+```
+
+In the call to `foo(2)` the compiler is allowed to simplify the expression `2 +
+x` down to `4`, but `4` must be materialized, for example, by storing it into
+memory, a register, etc., even though `4` is not read by anything afterwards
+because `black_box` could try to read it.
+
+### Benchmarking `Vec::push`
+
+The `hint::black_box` is useful for producing synthetic benchmarks that more
+accurately represent the behavior of a real application. In the following
+snippet, the function `bench` executes `Vec::push` 4 times in a loop:
+
+```rust
+fn push_cap(v: &mut Vec<i32>) {
+    for i in 0..4 {
+      v.push(i);
+    }
+}
+
+pub fn bench_push() -> Duration { 
+    let mut v = Vec::with_capacity(4);
+    let now = Instant::now();
+    push_cap(&mut v);
+    now.elapsed()
+}
+```
+
+Here, we allocate the `Vec`, push into it without growing its capacity, and drop
+it, without ever using it for anything. If we look at the assembly
+(https://rust.godbolt.org/z/wDckJF):
+
+
+```asm
+example::bench_push:
+  sub rsp, 24
+  call std::time::Instant::now@PLT
+  mov qword ptr [rsp + 8], rax
+  mov qword ptr [rsp + 16], rdx
+  lea rdi, [rsp + 8]
+  call std::time::Instant::elapsed@PLT
+  add rsp, 24
+  ret
+```
+
+it is pretty amazing: LLVM has actually managed to completely optimize the `Vec`
+allocation and call to `push_cap` away! In our real application, we would
+probably use the vector for something, preventing all of these optimizations
+from triggering, but in this synthetic benchmark LLVM optimizations are
+producing a benchmark that won't tell us anything about the cost of `Vec::push`.
+
+We can use `hint::black_box` to create a more realistic synthetic benchmark
+(https://rust.godbolt.org/z/CeXmxN):
```

Comment:
Please explain *why* this makes it more realistic. Namely: Because the compiler has to assume that `black_box` observes and mutates `v`, it cannot just optimize all this code away.